### PR TITLE
Update all submodules that rustbuild doesn't depend on lazily

### PR DESCRIPTION
**Pull request by jyn514**
This only updates the submodules the first time they're needed, instead
of unconditionally the first time you run x.py.

Ideally, this would move *all* submodules to rustbuild and not exclude some tools and
backtrace. Unfortunately, cargo requires all `Cargo.toml` files in the
whole workspace to be present to build any crate.

On my machine, this takes the time for an initial submodule clone (for
`x.py --help`) from 55.70 to 15.87 seconds.

Helps with https://github.com/rust-lang/rust/issues/76653. Builds on https://github.com/rust-lang/rust/pull/86015 and should not be merged before (only the last commit is relevant).

### CODE REVIEW

File: src/bootstrap/tool.rs [hunk 1]
```diff
@@ -686,21 +688,21 @@ macro_rules! tool_extended {
...
             compiler: self.compiler,
             target: self.target,
             extra_features: Vec::new(),
         });
         self.extra_features.push("clippy".to_owned());
     };
-    Rustfmt, rustfmt, "src/tools/rustfmt", "rustfmt", stable=true, {};
-    RustAnalyzer, rust_analyzer, "src/tools/rust-analyzer/crates/rust-analyzer", "rust-analyzer", stable=false, {};
+    Rustfmt, rustfmt, "src/tools/rustfmt", "rustfmt", stable=true, submodule="rustfmt", {};
+    RustAnalyzer, rust_analyzer, "src/tools/rust-analyzer/crates/rust-analyzer", "rust-analyzer", stable=false, submodule="rust-analyzer", {};
```
Comment (jyn514) [inline src/bootstrap/tool.rs:705]: For now, these are redundant because they're also updated in bootstrap.py; I added them before I realized Cargo depended on the whole workspace being checked out. Let me know if I should drop this change.

Comment (Mark-Simulacrum) [reply src/bootstrap/tool.rs:705]: Yeah, it's probably best to avoid redundant updates to the submodules necessary for Cargo to run.

Comment (jyn514) [reply src/bootstrap/tool.rs:705]: Sure thing, I changed this to only manage `rust-analyzer`.

**Comment by jyn514** [comment]
This is mostly working, but it now tries to update stdarch 3 times in a row, which is slow especially on windows.

```
$ x check
Updating only changed submodules
Submodules updated in 0.17 seconds
    Finished dev [unoptimized + debuginfo] target(s) in 0.69s
Updating submodule library/stdarch
Updating submodule src/doc/embedded-book
Updating submodule src/tools/rust-analyzer
Submodule path 'src/tools/rust-analyzer': checked out 'b82458818d44dfe5b4b5db38d8113e3f3194506e'
Updating submodule library\stdarch
Updating submodule library\stdarch
Checking stage0 std artifacts (x86_64-pc-windows-msvc -> x86_64-pc-windows-msvc)
```

**Comment by jyn514** [comment]
> This is mostly working, but it now tries to update stdarch 3 times in a row, which is slow especially on windows.


cc https://rust-lang.zulipchat.com/#narrow/stream/182449-t-compiler.2Fhelp/topic/x.2Epy.20always.20updates.20LLVM.20submodule

**Comment by petrochenkov** [comment]
@jyn514 
Can the updates be skipped only for submodules that are not yet initialized?

For submodules that are already initialized this results in a dirty git tree with out-of-date submodules - this is what I got after the recent LLVM changes.
Then you have to update manually with `git submodule update` instead of just running any `x.py` command which you are going to run anyway.
Partial updates usually require fetching only a couple of commits (except for perhaps LLVM version updates) and are therefore quick.

**Comment by jyn514** [comment]
> For submodules that are already initialized this results in a dirty git tree with out-of-date submodules - this is what I got after the recent LLVM changes.

Hmm, I can't replicate this behavior:

```
$ x check
Updating only changed submodules
Submodules updated in 0.01 seconds
    Finished dev [unoptimized + debuginfo] target(s) in 0.09s
Updating submodule src/llvm-project
Submodule path 'src/llvm-project': checked out '5f67a5715771b7d29e4713e8d68338602d216dcf'
Checking stage0 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
```

Do you remember the command you ran?

**Comment by jyn514** [comment]
No longer blocked since https://github.com/rust-lang/rust/pull/86015 :)

File: src/bootstrap/doc.rs [hunk 2]
```diff
@@ -60,12 +65,12 @@ macro_rules! book {
...
-    CargoBook, "src/tools/cargo/src/doc", "cargo";
-    EditionGuide, "src/doc/edition-guide", "edition-guide";
-    EmbeddedBook, "src/doc/embedded-book", "embedded-book";
-    Nomicon, "src/doc/nomicon", "nomicon";
-    Reference, "src/doc/reference", "reference";
-    RustByExample, "src/doc/rust-by-example", "rust-by-example";
+    CargoBook, "src/tools/cargo/src/doc", "cargo", submodule = ["src", "tools", "cargo"];
+    EditionGuide, "src/doc/edition-guide", "edition-guide", submodule = ["src", "doc", "edition-guide"];
+    EmbeddedBook, "src/doc/embedded-book", "embedded-book", submodule = ["src", "doc", "embedded-book"];
+    Nomicon, "src/doc/nomicon", "nomicon", submodule = ["src", "doc", "nomicon"];
```
Comment (Mark-Simulacrum) [inline src/bootstrap/doc.rs:71]: Hm, I'm a little surprised that the submodules here need to be arrays and join'd together given that src/doc/nomicon seems to work OK and I think in some other places we have forward-slash paths too. Did you observe problems from that?

File: src/bootstrap/lib.rs [hunk 3]
```diff
@@ -477,17 +477,110 @@ impl Build {
...
+            "src/tools/rls",
+            "src/tools/miri",
+            "library/backtrace",
+            "library/stdarch",
+        ];
+        let output = output(
+            Command::new("git")
+                .args(&["config", "--file"])
+                .arg(&self.config.src.join(".gitmodules"))
+                .args(&["--get-regexp", "path"]),
```
Comment (Mark-Simulacrum) [inline src/bootstrap/lib.rs:564]: Can we add some comment here about this looking for submodule.$name.path = $path in the configuration?

**Comment by Mark-Simulacrum** [comment]
A little worried about regressions here but overall looks pretty good. Left some comments.

File: src/bootstrap/doc.rs (hunk 2 above)
Comment (jyn514) [reply src/bootstrap/doc.rs:71]: No ... I guess this is premature bug avoidance, I was trying to prevent bugs on windows since I can't test there. But since other things are working fine it's probably overkill.

File: src/bootstrap/lib.rs (hunk 3 above)
Comment (jyn514) [reply src/bootstrap/lib.rs:564]: Sure, done.

**Comment by jyn514** [comment]
> A little worried about regressions here but overall looks pretty good. Left some comments.

I'm less worried about this than I was about https://github.com/rust-lang/rust/pull/81601 - this uses exactly the same logic, which has been merged to master for a week without any complaints :)

**Comment by jyn514** [comment]
I'm not sure what issue petrochenkov ran into before, but it seems to be fixed now:
```
$ x fmt
Updating only changed submodules
Submodules updated in 0.01 seconds
    Finished dev [unoptimized + debuginfo] target(s) in 26.83s
Updating submodule src/doc/nomicon
remote: Enumerating objects: 200, done.
remote: Counting objects: 100% (200/200), done.
remote: Compressing objects: 100% (80/80), done.
remote: Total 200 (delta 134), reused 175 (delta 119), pack-reused 0
Receiving objects: 100% (200/200), 92.45 KiB | 1.74 MiB/s, done.
Resolving deltas: 100% (134/134), completed with 27 local objects.
From https://github.com/rust-lang/nomicon
   55de6fa..ca4aa39  master     -> origin/master
Submodule path 'src/doc/nomicon': checked out 'b9ca313e687c991223e23e5520529815dc281205'
```
In particular note that this is being checked out by rustbuild, even though it's not one of the repos hard-coded to always be checked out - it noticed that I already have the submodule locally and it wasn't up to date.

File: src/bootstrap/doc.rs [hunk 4]
```diff
@@ -60,12 +73,12 @@ macro_rules! book {
 // NOTE: When adding a book here, make sure to ALSO build the book by
 // adding a build step in `src/bootstrap/builder.rs`!
```
Comment (Mark-Simulacrum) [inline src/bootstrap/doc.rs:74]: Can we add a note here about making sure submodule is added when populating these? I wonder if it's possible to automate the submodule detection -- in theory, we can at startup or otherwise have a list of submodules and do a prefix search or so (e.g., src/tools/cargo/src/doc will note that src/tools/cargo is a prefix and a submodule)...

Probably just leaving a comment is OK for now. We don't really add submodules much these days.

File: src/bootstrap/tool.rs [hunk 5]
```diff
@@ -736,6 +738,8 @@ macro_rules! tool_extended {
 
 // Note: tools need to be also added to `Builder::get_step_descriptions` in `builder.rs`
 // to make `./x.py build <tool>` work.
+// Note: Most submodule updates for tools are handled by bootstrap.py, since they're needed just to
+// build bootstrap.
```
Comment (Mark-Simulacrum) [inline src/bootstrap/tool.rs:742]: Can we rephrase this to something like "invoke Cargo"? In particular, no submodules are *actually* required to build bootstrap, and we could avoid the connection to bootstrap by creating a separate workspace for it. But, that wouldn't help with the need for these to get init'd anyway since they're still needed for any cargo invocation pretty much.

Comment (jyn514) [reply src/bootstrap/tool.rs:742]: > But, that wouldn't help with the need for these to get init'd anyway since they're still needed for any cargo invocation pretty much.

We could avoid cloning most of the tool subdirectories if we kept the compiler in a separate workspace from the standard library. cc https://github.com/rust-lang/rust/pull/76533

Anyway, I updated the comment.

**Comment by Mark-Simulacrum** [comment]
r=me but as before, would like to do so when you have a chance to monitor for issues that arise so we can revert/patch if needed.

**Comment by jyn514** [comment]
I think I also want to fix https://github.com/rust-lang/rust/issues/86954 before merging this; I'll try to do that tomorrow.

**Comment by jyn514** [comment]
Oh uhh apparently I forgot to ever merge this 🙃 I should have time tomorrow to revert if need be.

@bors r=Mark-Simulacrum rollup=iffy

**Comment by semarie** [comment]
While building -nightly **from tarball**, the build fails with:

```
running: /data/semarie/build-rust/build_dir/build/bootstrap/debug/bootstrap dist --jobs=4
warning: x.py has made several changes recently you may want to look at
help: consider looking at the changes in `src/bootstrap/CHANGELOG.md`
note: to silence this warning, add `changelog-seen = 2` at the top of `config.toml`
finding compilers
CC_x86_64-unknown-openbsd = "cc"
CFLAGS_x86_64-unknown-openbsd = ["-O2", "-ffunction-sections", "-fdata-sections", "-fPIC", "-m64", "-O2", "-pipe"]
CXX_x86_64-unknown-openbsd = "c++"
CXXFLAGS_x86_64-unknown-openbsd = ["-O2", "-ffunction-sections", "-fdata-sections", "-fPIC", "-m64", "-O2", "-pipe"]
AR_x86_64-unknown-openbsd = "ar"
running sanity check
learning about cargo
thread 'main' panicked at 'command did not execute successfully: "git" "config" "--file" "/data/semarie/build-rust/build_dir/rustc-nightly-src/.gitmodules" "--get-regexp" "path"
expected success, got: exit status: 1', src/bootstrap/lib.rs:568:22
stack backtrace:
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
Traceback (most recent call last):
  File "/data/semarie/build-rust/build_dir/rustc-nightly-src/x.py", line 27, in <module>
    bootstrap.main()
  File "/data/semarie/build-rust/build_dir/rustc-nightly-src/src/bootstrap/bootstrap.py", line 1209, in main
    bootstrap(help_triggered)
  File "/data/semarie/build-rust/build_dir/rustc-nightly-src/src/bootstrap/bootstrap.py", line 1195, in bootstrap
    run(args, env=env, verbose=build.verbose, is_bootstrap=True)
  File "/data/semarie/build-rust/build_dir/rustc-nightly-src/src/bootstrap/bootstrap.py", line 153, in run
    raise RuntimeError(err)
RuntimeError: failed to run: /data/semarie/build-rust/build_dir/build/bootstrap/debug/bootstrap dist --jobs=4
Thu Jul 22 18:07:30 CEST 2021: task not finished: see build.log for detail
```

it seems the build tries to run `git config --file path/to/.gitmodules --get-regexp path` whereas the tarball doesn't contains `.gitmodules` file.

**Comment by jyn514** [comment]
I have to look into how tarballs work - is there a reason they don't set `submodules = false`?

**Comment by Mark-Simulacrum** [comment]
We don't put any new default config in the tarballs (and I don't think we should); I imagine the right check here is something like is_git() and if so defaulting submodules to false.

**Comment by petrochenkov** [comment]
@jyn514 
As I previously complained in https://github.com/rust-lang/rust/pull/82653#issuecomment-849013698 this PR results in having a bunch of out-of-date submodules that are no longer synchronized on any `x.py` run, but require manually running `git submodule update`.

How to reproduce:
- Have some moderately old (e.g. two weeks) branch you've been previously working on as a current branch.
   ```sh
    $ git status
    On branch my_old_branch
    ```
- Make a fresh branch from `upstream/master`
    ```sh
    git fetch --all --prune
    git checkout -b my_fresh_branch upstream/master
    ```
- Run `x.py` to prepare for compiler work and watch youtube for some time
    ```sh
    ./x.py test src/test/ui --test-args zzzzzz
    ```
- Have a bunch of dirty submodules that pollute `git gui` and `git status` even after running `x.py`
   ```
	$ git status
	On branch my_fresh_branch
	Your branch is up to date with 'upstream/master'.

	Changes not staged for commit:
	  (use "git add <file>..." to update what will be committed)
	  (use "git restore <file>..." to discard changes in working directory)
			modified:   src/doc/book (new commits)
			modified:   src/doc/edition-guide (new commits)
			modified:   src/doc/embedded-book (new commits)
			modified:   src/doc/reference (new commits)
			modified:   src/doc/rust-by-example (new commits)
			modified:   src/doc/rustc-dev-guide (new commits)
			modified:   src/tools/rust-analyzer (new commits)
   ```

If the submodule was already initialized, then updating it very quick and can be done unconditionally, even if the given submodule is not strictly necessary for the given `x.py` command.
`x.py` did the right thing previously.
I agree that not cloning/initializing submodules that were never initialized previously is a viable optimization, but not updating already initializing submodules only results in inconveniences.

**Comment by jyn514** [comment]
> If the submodule was already initialized, then updating it is very quick and can be done unconditionally, even if the given submodule is not strictly necessary for the given x.py command.
x.py did the right thing previously.
I agree that not cloning/initializing submodules that were never initialized previously is a viable optimization, but not updating already initializing submodules only results in inconveniences.

Contrary to what you seem to believe, I'm not trying to break your workflow. Not updating the submodules is just a bug, it's not intentional. I think https://github.com/rust-lang/rust/pull/87443 should fix it.
